### PR TITLE
ENH: Formula-Viewer

### PR DESCRIPTION
**Pull request by aksharsarvesh**
Added Formulas. This should be the true PR, I'm going to archive the old one because it's on the wrong branch, and I've since been updating this branch

### CODE REVIEW

File: archive_viewer/mixins/traces_table.py [hunk 1]
```diff
@@ -162,73 +169,96 @@ def import_csv(self) -> None:
...
         self.setWindowTitle("Formula Input")
 
         # Create the layout for the dialog
         layout = QVBoxLayout(self)
-
         # Create the QLineEdit for formula input
         self.field = QLineEdit(self)
+        self.curveModel = self.parent().parent().curves_model
+        self.pv_list = QTableView(self)
+        #We're going to copy the list of PVs from the curve model. We're also not going to allow the user to make edits to the list of PVs
```
Comment (YektaY) [inline archive_viewer/mixins/traces_table.py:184]:
```suggestion
        # We're going to copy the list of PVs from the curve model. We're also not going to allow the user to make edits to the list of PVs
```

File: archive_viewer/mixins/traces_table.py [hunk 2]
```diff
@@ -162,73 +169,96 @@ def import_csv(self) -> None:
...
+            row = i // 6
+            col = i % 6
             grid_layout.addWidget(button, row, col)
-
             # Connect the button clicked signal to the appropriate action
+            # PV currently does nothing, this is a remnant
+            # From when we would have the pv_list open in a new window
             if button_text == "PV":
-                button.clicked.connect(lambda _: self.field.insert("PV"))
+                # TODO: Either give a function for this button or replace it
```
Comment (YektaY) [inline archive_viewer/mixins/traces_table.py:221]: if you don't think you need the PV button I would remove it

File: archive_viewer/mixins/traces_table.py [hunk 3]
```diff
@@ -162,73 +169,96 @@ def import_csv(self) -> None:
...
-        #   if yes, replace with row header; if no, add to TableModel and replace with row header
-        formula = self.field.text()
-        try:
-            result = eval(formula)
-            self.field.setText(str(result))
-        except (SyntaxError, TypeError):
-            self.field.setText("Error")
-            logger.error("Invalid formula evaluated.")
+    def keyPressEvent(self, e: QKeyEvent) -> None:
+        # Special key press tracker, just so that if enter or return is pressed the formula dialog attempts to submit the formula
```
Comment (YektaY) [inline archive_viewer/mixins/traces_table.py:235]: Style: I would suggest changing the comment into a docstring for the method

File: archive_viewer/mixins/traces_table.py [hunk 4]
```diff
@@ -162,73 +169,96 @@ def import_csv(self) -> None:
...
-        formula = self.field.text()
-        pv_name = self.pv_name_input.text()
-
-        print("Formula:", formula)
-        print("PV Name:", pv_name)
-
-        self.accept()
+        # We take in the formula (prepend the formula tag) and attempt to create a curve. Iff it passes, we close the window
+        formula = "f://" + self.field.text()
+        # pv_name = self.pv_name_input.text()
```
Comment (YektaY) [inline archive_viewer/mixins/traces_table.py:260]:
```suggestion

```

File: archive_viewer/mixins/traces_table.py [hunk 5]
```diff
@@ -162,73 +169,96 @@ def import_csv(self) -> None:
...
+            self.pv_list.setRowHidden(i, False)
+        index = self.curveModel.index(self.index.row(), 0)
+        curve = self.curveModel._plot._curves[self.index.row()]
+        if index.data() and isinstance(curve, FormulaCurveItem):
+            self.field.setText(str(index.data()).strip("f://"))
+        else:
+            self.field.setText("")
+        super().exec_()
 
     def accept_formula(self, **kwargs: Dict[str, Any]) -> None:
```
Comment (YektaY) [inline archive_viewer/mixins/traces_table.py:256]: Style: Would also change these comments into a docstring for the method

File: archive_viewer/table_models/curve_model.py [hunk 6]
```diff
@@ -107,14 +135,15 @@ def append(self, address: Optional[str] = None, name: Optional[str] = None, colo
...
-        if self.rowCount() != 1:
+        if self.rowCount() == 0:
             self._axis_model.append()
         y_axis = self._axis_model.get_axis(-1)
         if not color:
             color = ColorButton.index_color(self.rowCount())
         self._row_names.append(self.next_header())
         #          KLYS:LI22:31:KVAC
         self.beginInsertRows(QModelIndex(), len(self._plot._curves), len(self._plot._curves))
+        #by default, add a blank archivePlotCurveItem such that there's an empty row to add PVs or formulas to.
```
Comment (YektaY) [inline archive_viewer/table_models/curve_model.py:146]:
```suggestion
        # By default, add a blank archivePlotCurveItem such that there's an empty row to add PVs or formulas to.
```

File: archive_viewer/table_models/curve_model.py [hunk 7]
```diff
@@ -145,6 +174,90 @@ def set_model_curves(self, curves: List[Dict]) -> None:
...
+        FormulaCurve.formula_invalid_signal.connect(partial(self.invalidFormula, header = rowName))
+        # Need to check if Formula is referencing a dead row
+        FormulaCurve.redrawCurve()
+        self.plot.removeItem(curve)
+        # Disconnect everything and delete it, create a new Formula with the dictionary of curve
+        [ch.disconnect() for ch in curve.channels() if ch]
+        del curve
+        return True
+
+    def invalidFormula(self, header):
```
Comment (YektaY) [inline archive_viewer/table_models/curve_model.py:239]: Style: can you add a docstring to this method?

File: archive_viewer/table_models/curve_model.py [hunk 8]
```diff
@@ -145,6 +174,90 @@ def set_model_curves(self, curves: List[Dict]) -> None:
...
 
+    def replaceToArchivePlot(self, curve: BasePlotCurveItem, index: QModelIndex, address: str, color: Optional[QColor] = None):
+        self.append(address=address, name=address, color=color)
+        self._plot._curves[index.row()] = self._plot._curves[-1]
+        self.beginRemoveRows(QModelIndex(), self.rowCount() - 1, self.rowCount() - 1)
+        self.plot._curves = self.plot._curves[:-1]
+        self.plot.removeItem(curve)
+        self.endRemoveRows()
+
+    def formulaToPVDict(self, rowName: str, formula: str) -> dict:
```
Comment (YektaY) [inline archive_viewer/table_models/curve_model.py:185]: Style: can you add a docstring to this method?

File: archive_viewer/table_models/curve_model.py [hunk 9]
```diff
@@ -145,6 +174,90 @@ def set_model_curves(self, curves: List[Dict]) -> None:
         self.endResetModel()
         logger.debug("Finished setting curves model")
 
+    def replaceToArchivePlot(self, curve: BasePlotCurveItem, index: QModelIndex, address: str, color: Optional[QColor] = None):
```
Comment (YektaY) [inline archive_viewer/table_models/curve_model.py:177]: Style: can you add a docstring to this method?

File: archive_viewer/table_models/curve_model.py [hunk 10]
```diff
@@ -145,6 +174,90 @@ def set_model_curves(self, curves: List[Dict]) -> None:
...
+        for pv in pvs:
+            # Check if all of the requested rows actually exist
+            if pv not in self._row_names:
+                raise ValueError(f"{pv} is an invalid variable name")
+            elif pv == rowName:
+                raise ValueError(f"{pv} is recursive")
+            elif self._row_names.index(pv) > self._row_names.index(rowName):
+                raise ValueError(f"Error, all referenced curves must come before the Formula")
+            else:
+                # if it's good, add it to the dictionary of curves. rindex = row index (int) as opposed to index, which is a QModelIndex
```
Comment (YektaY) [inline archive_viewer/table_models/curve_model.py:197]:
```suggestion
                # If it's good, add it to the dictionary of curves. rindex = row index (int) as opposed to index, which is a QModelIndex
```

File: archive_viewer/table_models/curve_model.py [hunk 11]
```diff
@@ -145,6 +174,90 @@ def set_model_curves(self, curves: List[Dict]) -> None:
...
+        #          KLYS:LI22:31:KVAC
+        # Handle Archives and formulas differently
+        if index.row() == self.rowCount() - 1:
+            self.append()
+        y_axis = self._axis_model.get_axis(-1)
+        FormulaCurve = FormulaCurveItem(formula=formula, name=formula, pvs=pvdict, color=color, yAxisName=y_axis.name)
+        self._plot._curves[index.row()] = FormulaCurve
+        FormulaCurve.formula_invalid_signal.connect(partial(self.invalidFormula, header = rowName))
+        # Need to check if Formula is referencing a dead row
+        FormulaCurve.redrawCurve()
```
Comment (zdomke) [inline archive_viewer/table_models/curve_model.py:232]: Not sure if this error belongs here or in https://github.com/slaclab/pydm/pull/1095
When setting the formula to blank text, whether by typing "f://" as the channel or open FormulaDialog and click "OK", a syntax error gets thrown here by the `redrawCurve()`.

Here is the full traceback:
```
Traceback (most recent call last):
  File "/afs/slac.stanford.edu/u/cd/zdomke/workspace/PyDMArchiveViewer/Archive-Viewer/archive_viewer/mixins/traces_table.py", line 261, in accept_formula
    passed = self.curveModel.replaceToFormula(index = self.curveModel.index(self.parent().selected_index.row(), 0), formula = formula)
  File "/afs/slac.stanford.edu/u/cd/zdomke/workspace/PyDMArchiveViewer/Archive-Viewer/archive_viewer/table_models/curve_model.py", line 232, in replaceToFormula
    FormulaCurve.redrawCurve()
  File "/afs/slac.stanford.edu/u/cd/zdomke/workspace/pydm/pydm/widgets/archiver_time_plot.py", line 553, in redrawCurve
    y = [eval(self._trueFormula), eval(self._trueFormula)]
  File "<string>", line 0

SyntaxError: invalid syntax
```

Comment (zdomke) [reply archive_viewer/table_models/curve_model.py:232]: idk why someone would want to do this, but it is a strange edge case

**Review by zdomke — COMMENTED** [review]
It's looking pretty good. I mostly have styling suggestions like Yekta.

I did manage to get an error to pop, but I'm not sure if the fix belongs in Archive-Viewer or PyDM

File: archive_viewer/mixins/traces_table.py (hunk 2 above)
Comment (aksharsarvesh) [reply archive_viewer/mixins/traces_table.py:221]: I did not consider this but I definitely could add it. No, it never added/hit the list of PVs. you may be thinking of the PV search tool which is one of the options in the custom dialogue box, one level above the formula dialogue box

Comment (zdomke) [reply archive_viewer/mixins/traces_table.py:221]: Weird. I don't know what I'm thinking of then. I really thought that was an idea you had messed around with and implemented. Oh well.

Not sure it's the most useful functionality for this button, could just cause confusion. Not sure what this button could do.
Is there a more useful function that we could include instead? Even just `^e`? Might need to move the button to a new place too

File: archive_viewer/table_models/curve_model.py (hunk 11 above)
Comment (aksharsarvesh) [reply archive_viewer/table_models/curve_model.py:272]: Fixed

**Review by YektaY — APPROVED** [review]

File: archive_viewer/mixins/traces_table.py [hunk 12]
```diff
@@ -199,8 +198,9 @@ def __init__(self, parent: QObject) -> None:
             # PV currently does nothing, this is a remnant
             # From when we would have the pv_list open in a new window
             if button_text == "PV":
-                # TODO: Either give a function for this button or replace it
-                pass
+                self.PVButton = button
+                self.PVButton.setCheckable(True)
+                self.PVButton.clicked.connect(self.showPVList)
```
Comment (zdomke) [inline archive_viewer/mixins/traces_table.py:203]: The PVButton is unchecked by default, so the PV list is hidden by default. Is this intended?
If so, I think it makes more sense to show it by default.
```suggestion
                self.PVButton = button
                self.PVButton.setCheckable(True)
                self.PVButton.setChecked(True)
                self.PVButton.clicked.connect(self.showPVList)
```

**Review by zdomke — COMMENTED** [review]
Looks like it's there. I just noticed that there is still a print statement. Also the formula dialog box doesn't show by default. Unsure if that's intended.

File: archive_viewer/mixins/traces_table.py (hunk 12 above)
Comment (aksharsarvesh) [reply archive_viewer/mixins/traces_table.py:203]: That was intentional, and sounds good I'll change it

**Comment by aksharsarvesh** [comment]
> Looks like it's there. I just noticed that there is still a print statement. Also the formula dialog box doesn't show by default. Unsure if that's intended.

Formula Dialog? Do you mean the PV list in the formula dialog (your other comment). On my end Formula Dialog opens as usual

**Comment by zdomke** [comment]
> > Looks like it's there. I just noticed that there is still a print statement. Also the formula dialog box doesn't show by default. Unsure if that's intended.
> 
> Formula Dialog? Do you mean the PV list in the formula dialog (your other comment). On my end Formula Dialog opens as usual

My mistake, I did mean the PV List

**Review by zdomke — APPROVED** [review]

**Pull request merged**